src/typedefs.jl

```diff
@@ -73,7 +75,7 @@ end
 
 ## Nullable Arrays
 
-@compat abstract type AbstractNullableCategoricalArray{T, N, R} <: AbstractArray{Nullable{CategoricalValue{T, R}}, N} end
+@compat abstract type AbstractNullableCategoricalArray{T, N, R} <: AbstractArray{Union{CategoricalValue{T, R}, Null}, N} end
```

Comment:
Actually the only solution I could find is to have
```julia
abstract type AbstractCategoricalArray{T, N, R, V, U} <: AbstractArray{Union{CategoricalValue{V, R}, U}, N} end
immutable CategoricalArray{T, N, R <: Integer, V, U} <: AbstractCategoricalArray{T, N, R, V, U} [...] end
```
Then a non-nullable `CategoricalArray` will look like `CategoricalArray{String, 1, UInt32, String, Union{}}`, and `NullableCategoricalArray` will be replaced with `CategoricalArray{Union{String, Null}, 1, UInt32, String, Null}`.

I'm not really happy about the need for the two last parameters, which are redundant with `T`, but they are required (AFAICT) to define the type of the pool and the element type when subtyping `AbstractArray`. At least they are mostly invisible to the user, since the inner constructor computes them automatically and they are not needed for dispatch (we could even avoid printing them).

I can push a mostly working version of this later today if you're interested.

Regarding `WeakRefStringArray`, I'm not sure how it relates to categorical arrays. Do you want it to be a subtype of `AbstractCategoricalArray`?